src/monty/io.py

```diff
@@ -54,7 +55,53 @@ def zopen(filename: Union[str, Path], *args, **kwargs) -> IO:
     return open(filename, *args, **kwargs)
 
 
-def reverse_readfile(filename: Union[str, Path]) -> Generator[str, str, None]:
+def _get_line_ending(
+    file: str | Path | io.TextIOWrapper,
+) -> Literal["\r\n", "\n", "\r"]:
+    """Helper function to get line ending of a file.
+
+    This function assumes the file has a single consistent line ending.
+
+    Returns:
+        "\n": Unix line ending.
+        "\r\n": Windows line ending.
+        "\r": Classic MacOS line ending.
+
+    Raises:
+        ValueError: If line ending is unknown.
+
+    Warnings:
+        If file is empty, "\n" would be used as default.
+    """
+    if isinstance(file, (str, Path)):
+        with open(file, "rb") as f:
+            first_line = f.readline()
+    elif isinstance(file, io.TextIOWrapper):
+        first_line = file.buffer.readline()
+    elif isinstance(file, (gzip.GzipFile, bz2.BZ2File)):
+        first_line = file.readline()
+        file.seek(0)  # reset pointer
+    else:
+        raise TypeError(f"Unknown file type {type(file).__name__}")
+
+    # Return Unix "\n" line ending as default if file is empty
+    if not first_line:
+        warnings.warn("File empty, use default line ending \n.", stacklevel=2)
+        return "\n"
+
+    if first_line.endswith(b"\r\n"):
+        return "\r\n"
+    if first_line.endswith(b"\n"):
+        return "\n"
+    if first_line.endswith(b"\r"):
+        return "\r"
+
+    raise ValueError(f"Unknown line ending in line {repr(first_line)}.")
+
+
+def reverse_readfile(
```

Comment:
Current implementation has another issue, say we have a line `repr` as `"line1\nline2\nline3\n"`, when read reversely, it becomes four lines because we just identify lines by finding `\n` (`filemap[line_end_pos + len(l_end) : file_size]`).